Rapport de QA
--------------------

Tout d'abord, avant de faire la QA, j'essaie toujours de reproduire le bug à l'origine de la PR. Je le précise, parce que j'ai quelques remarques. Si on fait `make test-back`, je n'ai pas le faux positif signalé. En effet, si je lance:
```
python manage.py test --settings zds.settings.test zds.utils.tests.test_misc.Misc.test_remove_url_scheme 
```
Le test passe.

En revanche, si j'utilise les paramètres `dev`:
```
 python manage.py test --settings zds.settings.dev zds.utils.tests.test_misc.Misc.test_remove_url_scheme
```
en effet, ça plante:
```
FAIL: test_remove_url_scheme (zds.utils.tests.test_misc.Misc)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/media/BigData/linux/Développement/zds/zds-site/zds/utils/tests/test_misc.py", line 57, in test_remove_url_scheme
    self.assertEqual(remove_url_scheme(element.given), element.expected)
AssertionError: '127.0.0.1:8000/media/gallery/1/1.png' != '/media/gallery/1/1.png'
- 127.0.0.1:8000/media/gallery/1/1.png
? --------------
+ /media/gallery/1/1.png
```

Maintenant, avec la PR, les deux façons de lancer les tests (avec les paramètres `test` ou `dev`) fonctionnent. Pas de bug lié à ce module sur la suite de tests complète (`make run-back`). Le site fonctionne correctement.

**QA OK** :heavy_check_mark: 

Un rebase et c'est bon.

--- a/zds/utils/templatetags/remove_url_scheme.py
+++ b/zds/utils/templatetags/remove_url_scheme.py
@@ -1,27 +1,37 @@
+import urllib.parse
+
 from django import template
 from django.conf import settings
-from six.moves import urllib_parse as urlparse
 
 register = template.Library()
 
 
 @register.filter('remove_url_scheme')
-def remove_url_scheme(input_url):
+def remove_url_scheme(url):
     """
-    make every image url pointing to this website protocol independant so that if we use https, we are sure
-    that all our media are served with this protocol.
+    Remove the scheme and hostname from a URL if it is internal, but leave it unchanged otherwise.
 
-    .. notice::
+    The internal hostname is determined using the value of ``ZDS_APP['site']['dns']``.
+    URLs with no scheme are accepted. URLs with no hostname are treated as internal.
 
-        this also removes the ``settings.ZDS_APP['site']['dns']`` from the url.
+    For example, ``http://zestedesavoir.com/media/gallery/1/1.png`` becomes ``/media/gallery/1/1.png``,
+    whereas ``/media/gallery/1/1.png`` and ``example.com/media/gallery/1/1.png`` stay the same.
 
-    :return: the url without its scheme, e.g. ``http://zestedesavoir.com/media/gallery/1/1.png`` becomes
-    ``/media/gallery/1/1.png``
-
+    :return: the url without its scheme and hostname.
     """
 
-    schemeless_url = input_url[len(urlparse.urlparse(input_url).scheme):]
-    schemeless_url = schemeless_url[len('://'):] if schemeless_url.startswith('://') else schemeless_url
-    if schemeless_url.startswith(settings.ZDS_APP['site']['dns']):
-        return schemeless_url[len(settings.ZDS_APP['site']['dns']):]
-    return input_url
+    # Parse URLs after adding a prefix if necessary (e.g 'zestedesavoir.com' becomes '//zestedesavoir.com')
+    url_normalized = url
+    if '//' not in url:
+        url_normalized = '//' + url
+    url_parsed = urllib.parse.urlsplit(url_normalized)
+
+    # Return external URLs unchanged
+    if url_parsed.netloc != settings.ZDS_APP['site']['dns']:
+        return url
+
+    # Clean internal URLs
+    url_noscheme = urllib.parse.urlunsplit(['', '', url_parsed.path, url_parsed.query, url_parsed.fragment])
+    url_cleaned = url_noscheme[0:]  # remove first "/"
+
+    return url_cleaned

--- a/zds/utils/tests/test_misc.py
+++ b/zds/utils/tests/test_misc.py
@@ -1,13 +1,10 @@
-import collections
 import datetime
 from django.test import TestCase
-from django.conf import settings
 from zds.member.factories import ProfileFactory, StaffProfileFactory
 from zds.tutorialv2.factories import PublishedContentFactory
 from zds.utils.misc import contains_utf8mb4
 from zds.utils.models import Alert
 from zds.utils.context_processor import get_header_notifications
-from zds.utils.templatetags.remove_url_scheme import remove_url_scheme
 
 
 class Misc(TestCase):
@@ -34,24 +31,3 @@ class Misc(TestCase):
         filter_result = get_header_notifications(staff.user)['alerts']
         self.assertEqual(1, filter_result['total'])
         self.assertEqual(alert.text, filter_result['list'][0]['text'])
-
-    def test_remove_url_scheme(self):
-        Element = collections.namedtuple('element', ['name', 'given', 'expected'])
-        oracle = {
-            Element('cannonical http', 'http://{}/media/gallery/1/1.png'.format(settings.ZDS_APP['site']['dns']),
-                    '/media/gallery/1/1.png'),
-            Element('cannonical no scheme internal', '{}/media/gallery/1/1.png'.format(settings.ZDS_APP['site']['dns']),
-                    '/media/gallery/1/1.png'),
-            Element('cannonical no scheme external', 'example.com/media/gallery/1/1.png',
-                    'example.com/media/gallery/1/1.png'),
-            Element('cannonical https', 'https://{}/media/gallery/1/1.png'.format(settings.ZDS_APP['site']['dns']),
-                    '/media/gallery/1/1.png'),
-            Element('limit: empty url', '',
-                    ''),
-            Element('old bug: url in qstring', 'http://example.com?q=http://{}'.format(settings.ZDS_APP['site']['dns']),
-                    'http://example.com?q=http://{}'.format(settings.ZDS_APP['site']['dns'])),
-        }
-
-        for element in oracle:
-            # as we are not in py3 we do not have subTest method. so we use a bare for loop.
-            self.assertEqual(remove_url_scheme(element.given), element.expected)

--- a/zds/utils/tests/tests_remove_url_scheme.py
+++ b/zds/utils/tests/tests_remove_url_scheme.py
@@ -1,17 +1,54 @@
+import copy
+
 from django.conf import settings
-from django.test import TestCase
+from django.test import TestCase, override_settings
 
 from zds.utils.templatetags.remove_url_scheme import remove_url_scheme
 
 
-class RemoveUrlProtocolTest(TestCase):
+class RemoveUrlSchemeTests(TestCase):
 
-    def test_remove_protocole_when_local_url(self):
-        self.assertEqual('/bla.html', remove_url_scheme('http://' + settings.ZDS_APP['site']['dns'] + '/bla.html'))
-        self.assertEqual('/bla.html', remove_url_scheme('https://' + settings.ZDS_APP['site']['dns'] + '/bla.html'))
+    @staticmethod
+    def get_cases(internal_hostname):
+        """Return test cases corresponding to different URLs types."""
+        return {
+            'no scheme, no hostname':
+                {'input': '/bla.html',
+                 'expected_output': '/bla.html'},
+            'http scheme, internal hostname':
+                {'input': 'http://{}/media/gallery/1/1.png'.format(internal_hostname),
+                 'expected_output': '/media/gallery/1/1.png'},
+            'https scheme, internal hostname':
+                {'input': 'https://{}/media/gallery/1/1.png'.format(internal_hostname),
+                 'expected_output': '/media/gallery/1/1.png'},
+            'no scheme, internal hostname':
+                {'input': '{}/media/gallery/1/1.png'.format(internal_hostname),
+                 'expected_output': '/media/gallery/1/1.png'},
+            'no scheme, external hostname':
+                {'input': 'example.com/media/gallery/1/1.png',
+                 'expected_output': 'example.com/media/gallery/1/1.png'},
+            'http scheme, external hostname, internal hostname in query':
+                {'input': 'http://example.com/?q=http://{}'.format(internal_hostname),
+                 'expected_output': 'http://example.com/?q=http://{}'.format(internal_hostname)}
+        }
 
-    def test_no_change_when_no_protocole(self):
-        self.assertEqual('/bla.html', remove_url_scheme('/bla.html'))
+    def run_cases(self, hostname):
+        """Test url_remove_scheme on all URL cases for a given hostname."""
+        overridden_zds_app = copy.deepcopy(settings.ZDS_APP)
+        overridden_zds_app['site']['dns'] = hostname
+        with override_settings(ZDS_APP=overridden_zds_app):
+            internal_hostname = settings.ZDS_APP['site']['dns']
+            test_cases = RemoveUrlSchemeTests.get_cases(internal_hostname)
+            for case_name, case in test_cases.items():
+                with self.subTest(msg=case_name):
+                    self.assertEqual(case['expected_output'], remove_url_scheme(case['input']))
 
-    def test_no_change_when_extern_address(self):
-        self.assertEqual('http://www.google.com/bla.html', remove_url_scheme('http://www.google.com/bla.html'))
+    def test_url_remove_scheme(self):
+        """Test url_remove_scheme for different hostnames."""
+        hostnames = ['127.0.0.1:8000',  # Raw IP with port
+                     'localhost:8000',  # Name with port
+                     'beta.zestedesavoir.com',  # Name without port, but with subdomain
+                     'zestedesavoir.com']  # Name without port again
+        for hostname in hostnames:
+            with self.subTest(msg=hostname):
+                self.run_cases(hostname)

